File: backend/src/main/kotlin/net/animeta/backend/controller/UserController.kt
package net.animeta.backend.controller

import net.animeta.backend.dto.UserDTO
import net.animeta.backend.exception.ApiException
import net.animeta.backend.model.User
import net.animeta.backend.repository.UserRepository
import net.animeta.backend.security.CurrentUser
import net.animeta.backend.serializer.UserSerializer
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/v2/users/{name}")
class UserController(val userRepository: UserRepository, val userSerializer: UserSerializer) {
    @GetMapping
    fun get(@PathVariable name: String, @CurrentUser(required = false) currentUser: User?): UserDTO {
        val user = userRepository.findByUsername(name) ?: throw ApiException.notFound()
        return userSerializer.serialize(user, currentUser)
    }
}

File: backend-ts/src/controllers/user.controller.ts
import { Controller, DefaultValuePipe, Get, Param, Query } from "@nestjs/common";
import { UserDTO } from 'shared/types';
import { CurrentUser } from "src/auth/decorators";
import { User } from "src/entities/user.entity";
import { UserSerializer, UserSerializerOptions } from "src/serializers/user.serializer";
import { InjectRepository } from "@nestjs/typeorm";
import { Repository } from "typeorm";
import { ApiException } from "./exceptions";

@Controller('/api/v4/users/:name')
export class UserController {
  constructor(
    @InjectRepository(User) private userRepository: Repository<User>,
    private userSerializer: UserSerializer,
  ) {}

  @Get()
  async get(
    @Param('name') name: string,
    @Query('options', new DefaultValuePipe('{}')) optionsJson: string,
    @CurrentUser({ required: false }) currentUser: User | null,
  ): Promise<UserDTO> {
    const user = await this.userRepository.findOne({ where: {username: name} })
    if (!user) throw ApiException.notFound()
    const options: UserSerializerOptions = JSON.parse(optionsJson)
    return this.userSerializer.serialize(user, currentUser, options)
  }
}
